apollo-router/src/plugin/mod.rs

```diff
@@ -48,10 +48,33 @@ use crate::RouterResponse;
 use crate::SubgraphRequest;
 use crate::SubgraphResponse;
 
-type InstanceFactory = fn(&serde_json::Value) -> BoxFuture<Result<Box<dyn DynPlugin>, BoxError>>;
+type InstanceFactory =
+    fn(&serde_json::Value, String) -> BoxFuture<Result<Box<dyn DynPlugin>, BoxError>>;
 
 type SchemaFactory = fn(&mut SchemaGenerator) -> schemars::schema::Schema;
 
+/// Initialise details for a plugin
+pub struct PluginInit<T> {
+    /// Configuration
+    pub config: T,
+    /// Router Supergraph Schema
+    pub schema: String,
+}
+
+impl<T> PluginInit<T>
+where
+    T: for<'de> Deserialize<'de>,
+{
+    pub fn new(config: T, schema: String) -> Self {
+        PluginInit { config, schema }
+    }
+
+    pub fn try_new(config: serde_json::Value, schema: String) -> Result<Self, BoxError> {
+        let config: T = serde_json::from_value(config)?;
+        Ok(PluginInit { config, schema })
+    }
```

Comment:
Could these be `pub(crate)` instead?

Reply:
`new` is used in the integration tests, so I can't make that `pub(crate)`. `try_new` is used in the new `supergraph_sdl` example. Arguably I could use `new` in the example, but it's more convenient to use `try_new`, so I think they both need to be pub.